lib/lsperfm/core/stats.rb

```diff
@@ -0,0 +1,64 @@
+# encoding: utf-8
+require "thread"
+
+Thread.abort_on_exception = true
+
+module LogStash::PerfM
+
+  class Stats
+
+    REFRESH_COUNT = 100
+
+    attr_accessor :stats
+
+    def initialize
+      @stats = []
+    end
+    # below stats counter and output reader threads are sharing state using
+    # the @stats_lock mutex, @stats_count and @stats. this is a bit messy and should be
+    # refactored into a proper class eventually
+
+    def detach_stats_counter
+      Thread.new do
+        loop do
+          start = @stats_lock.synchronize{@stats_count}
+          sleep(1)
+          @stats_lock.synchronize{@stats << (@stats_count - start)}
+        end
+      end
+    end
+
+    # detach_output_reader spawns a thread that will fill in the @stats instance var with tps samples for every seconds
+    # @stats access is synchronized using the @stats_lock mutex but can be safely used
+    # once the output reader thread is completed.
+    def detach_output_reader(io, regex)
+      Thread.new(io, regex) do |io, regex|
+        i = 0
+        @stats = []
+        @stats_count = 0
+        @stats_lock = Mutex.new
```

Comment:
I'm not familiar with this code, but if this detach_output_reader method is called more than once you're redefine the @stats_lock variable and ruin the locking for the other.
And on the other, if the only method that uses the lock is going to be invoked in here (detach_stats_counter), maybe it could be passed as an argument and avoid a @var?